Certificate based authorization

elcallio opened the pull request:
Fixes #10099
    
Adds the com.scylladb.auth.CertificateAuthenticator type. If set as authenticator, will extract roles from TLS authentication certificate (not wire cert - those are server side) subject, based on configurable regex.

Example:

scylla.yaml:

```
    authenticator: com.scylladb.auth.CertificateAuthenticator
    auth_superuser_name: <name>
    auth_certificate_role_query: CN=([^,\s]+)
    
    client_encryption_options:
      enabled: True
      certificate: <server cert>
      keyfile: <server key>
      truststore: <shared trust>
      require_client_auth: True
```
In a client, then use a certificate signed with the <shared trust> store as auth cert, with the common name <name>. I.e. for  qlsh set "usercert" and "userkey" to these certificate files.

No user/password needs to be sent, but role will be picked up from auth certificate. If none is present, the transport will reject the connection. If the certificate subject does not contain a recongnized role name (from config or set in tables) the authenticator mechanism will reject it.

Otherwise, connection becomes the role described.

To facilitate this, this also contains the addition of allowing setting super user name + salted passwd via command line/conf + some tweaks to SASL part of connection setup.

elcallio commented:
Note: no unit tests added yet. Working on it.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3368/

avikivity commented:
> based on configurable regex.
> 

Trigger word!

avikivity commented on auth/password_authenticator.cc:
```diff
@@ -106,13 +116,17 @@ future<> password_authenticator::migrate_legacy_metadata() const {
+            std::string salted_pwd(get_config_value(_qp.db().get_config().auth_superuser_salted_password(), ""));
+            if (salted_pwd.empty()) {
+                salted_pwd = passwords::hash(DEFAULT_USER_PASSWORD, rng_for_salt);
+            }
```
This is frightening, esp. in relation with unfixed #1796. After a topology change and before a repair, this acts as a password reset to a known password, no? Though I see the patch doesn't introduce the problem.

elcallio replied:
This only sets a default IFF there is no entry in auth tables already. If you have already initialized the system with either option of just a normal "create user..." thingy, it will do nothing. As stated in the config.cc section.

avikivity replied:
Lack of #1796 means that when adding a datacenter or when increasing the replication factor, there's a window (before repair) where queries return nothing.

avikivity commented on auth/roles-metadata.cc:
```diff
@@ -46,59 +46,42 @@ constexpr std::string_view qualified_name("system_auth.roles");
+            co_return p(results->one());
+        }
+    }
+    co_return false;
```
I can't see what changed due to concurrent coroutinization.

avikivity replied:
I still can't see it without replaying the coroutinization in my mind.

avikivity replied:
so yes, I prefer to see the coroutinization first.

elcallio replied:
It is there. There is an initial patch added: "[roles-metadata: Coroutinuze some helpers](https://github.com/scylladb/scylladb/pull/12214/commits/dbf0f4947debb59ce779542f17ad48285df5ce68)".

avikivity commented on auth/roles-metadata.cc:
```diff
@@ -46,59 +46,42 @@ constexpr std::string_view qualified_name("system_auth.roles");
+        auto superuser = rolename ? std::string_view(*rolename) : meta::DEFAULT_SUPERUSER_NAME;
+        const bool is_nondefault = row.get_as<sstring>(col_name) != superuser;
+        return is_nondefault && p(row);
     });
```
similar.

elcallio replied:
Same exact answer.

avikivity commented on db/config.cc:
```diff
@@ -909,6 +912,11 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , cache_index_pages(this, "cache_index_pages", liveness::LiveUpdate, value_status::Used, false,
         "Keep SSTable index pages in the global cache after a SSTable read. Expected to improve performance for workloads with big partitions, but may degrade performance for workloads with small partitions.")
+    , auth_superuser_name(this, "auth_superuser_name", value_status::Used, "", 
+        "Initial authentication super username. Ignored if authentication tables already contain a super user")
```
Why not default to cassandra?

elcallio replied:
No real reason other than a deep desire to move away from that default...

avikivity replied:
I think it's okay to put it there. Removes a helper, and easier for the users to determine what the default it.

elcallio replied:
My only argument against is that by this logic, the password option should have 'cassandra' as salted hash as default. However, this requires the deployment host to have said hash, which might require us to limit it to a less secure hash? (I would really suspect all expected platforms to have SHA512, but...). A very ugly default in any case... (Output of mkpasswd)

avikivity commented on db/config.cc:
```diff
@@ -909,6 +912,11 @@ db::config::config(std::shared_ptr<db::extensions> exts)
+        "Initial authentication super username. Ignored if authentication tables already contain a super user")
+    , auth_superuser_salted_password(this, "auth_superuser_salted_password", value_status::Used, "", 
+        "Initial authentication super user salted password. Create using mkpassword or similar. The hashing algorithm used must be available on the node host. "
+        "Ignored if authentication tables already contain a super user password.")
```
I'm not a fan of use-once options but don't have anything better to suggest (that preserves ease of use).

elcallio replied:
Driven by 
1.) The need to specify an initial user for cert auth
2.) Complaints about the cassandra:cassandra default previously, i.e. time during which an authenticating node runs with a known user/passwd. 
Same reservations about use-once here, but I have no better alternative (considered adding yet another config file, but that does not really change anything).

avikivity replied:
Oh I agree, just felt compelled to complain.

avikivity commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,152 @@
+        auto query = conf.auth_certificate_role_query();
+
+        if (query.empty()) {
+            throw std::invalid_argument("No role extraction query specified.");
```
Can we provide a reasonable default?

elcallio replied:
If someone suggests one. My best choice would be `CN=(\w+)`, i.e. CNAME = role. But I will let those that asked for the feature chime in before deciding anything.

avikivity replied:
@tzach @tnozicka

avikivity replied:
@tzach @tnozicka please comment on default regex to extract role from certificate.

avikivity replied:
btw, what about security? Can a user specify some other property (let's say city) as CN=superuser and pretend to be the superuser?

elcallio replied:
He can, but since we only accept the certificate subject iff the certificate signature itself verifies against the configured trust store, he would also have to steal/have the signing certificates used to issue the accepted certs. Which I guess is equivalent to knowing a normal password. But harder (?).

The only strengthening I can think of might be useful is perhaps restricting the allowed signers of super user certificates further, though that becomes tricky, both w.r. configuring and also querying the info from gnutls.

avikivity replied:
The attacker may send a certificate to be signed to the security administrator, and get it signed if the security administrator doesn't look at the other fields.

elcallio replied:
I assume certificates and id:s are not distributed like that. It would be the equivalent of handing out passwords on demand. The idea is for the admin (scylla serverless team) to assign an ID to a user/user group, sign a cert with the proper subject, and he then uses it.

But again, if we worry about specificly pretending to be super user, having a more complex trust setup (either subject mask for signing cert or separate trust store) for super user would at least prevent that. 
However, again, for a multi-tenant setup, pretending to be a different user (non-super user) would be just as bad (almost), so I think it is a problem that can't really be 100% solved. Signing authority need to be careful when issuing certs, Same as any other auth.

avikivity replied:
I don't know enough about how certificates are issued to say. What's the existing practice?

/cc @Lorak-mmk in case he knows.

elcallio replied:
I would assume it depends on issuer/org. But for this use case, I still assume that a central authority (sysadm) creates roles + cert + signs it with his root cert, then send the result to the client which will connect.

avikivity replied:
@tzach @tnozicka still need input here

elcallio replied:
Why the "certificate_property" stuff? A simple regex handles this (selecting different attributes). I.e. `auth_certificate_role_query: (?:CN=([^,\s]+))|(?:SAN=([^,\s]+))`.  (You even provide an example below)

elcallio replied:
SAN as source might be problematic. Not to mention a misuse of information, since it should be limited to DNS names...

elcallio replied:
... Well, email and other... I guess it could be valid. Starting to get your request on which field to query. Huge downside in that this requires changes to seastar first though, since we have no way to get alt names right now.

tnozicka replied:
regarding the cert safety, the client CA is preferably a dedicated one (that's why most people don't need the regex parsing as they control the CN used for signing) but even with the mapping both client CA and the mapping is controlled by an administrator a so a user can't ask to validate against a different fields to forfeit an identity. Usually it won't come to SANs either as there is not really a need for domains with dedicated client CA.

elcallio replied:
Well, you still need regex (or similar) to extract the relevant role name from a subject (which is a map of values, CN=... etc). While I am not sold on allowing SAN as well, the argument was that normal subject length is limited, and @Ten0  wants potentially longer role names (IIUC?). I can see the value in this, though it seems like a bit of a stretched issue... Just use shorter role names. :-) But I think his notion is using "real" certificates, not special, home-brewed auth-only ones, which might be a use case.

tnozicka replied:
>> Can we provide a reasonable default?
> My best choice would be CN=(\w+), i.e. CNAME = role

I agree this should but an advanced option and by default it should just use the CN

elcallio replied:
That is literally "just use the CN", since CN is an attribute in the subject string. Note that we are _not_ breaking the subject up into attribute pairs, we treat it as a fully matched string. CN=(\w) mean literally "take the (word-only) string of CN attribute).

Ten0 replied:
It's not that want longer role names, it's that I want to authenticate based on host names (e.g give role `my-service` to somebody who has a cert for `my-service--instance-12.prod.my-example-company.com`), and host names don't fit in CN.

elcallio replied:
Ah, ok, that kinda makes sense. :-)

avikivity commented:
Looks good, but where tests?

avikivity commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,152 @@
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
```
Our license is AGPL-3.0-or-later. Apache-2.0 is for code derived from Apache code.

elcallio replied:
fixed

avikivity commented:
> I am adding dtests for it. Coming soon. But anything to do with dtest + certs is painful.

How about test.py? It gained a lot of powers recently. @kostja can help.

elcallio commented:
There is a lot of infrastructure in dtest. Also, I don't see us having any infrastructure for manipulating config values, starting+stopping+restarting nodes etc in our pytest stuff...

kostja commented:
We have all of it @elcallio, please take a look at topology/ suite.

elcallio commented:
Not quite all. Some, but there are still boiler plate stuff needed to build some stuff. I am not super hot on trying to add those now, so at least initially a dtest will have to do. Will happily move it into our unit pytests once I feel more comfortable there. 

See link above for added dtest for features.

elcallio commented:
Fixed most of @avikivity remarks. Still not sure about the config default values though.

elcallio commented:
ping?

elcallio commented:
ping?

avikivity commented:
Sorry, reviewing.

avikivity commented on auth/roles-metadata.cc:
```diff
@@ -51,54 +51,36 @@ future<bool> default_role_row_satisfies(
-            return make_ready_future<bool>(p(results->one()));
-        });
-    });
+    for (auto c : { db::consistency_level::ONE, db::consistency_level::QUORUM }) {
```
Unrelated: I think we should switch the order. First try quorum, then fall back to ONE on failure (not on empty rows).

avikivity commented on auth/roles-metadata.cc:
```diff
@@ -51,54 +51,36 @@ future<bool> default_role_row_satisfies(
-        });
-    });
+    for (auto c : { db::consistency_level::ONE, db::consistency_level::QUORUM }) {
+        auto results = co_await qp.execute_internal(query, db::consistency_level::ONE
```
Looks like you tried to concurrently change the code layout to use a loop, but forgot to use the loop variable here.

elcallio replied:
Fixed.

avikivity commented on auth/password_authenticator.cc:
```diff
@@ -161,6 +175,8 @@ future<> password_authenticator::stop() {
 
 db::consistency_level password_authenticator::consistency_for_user(std::string_view role_name) {
+    // TODO: this is plain dung. Why treat hardcoded default special, but for example a user-created
+    // super user uses plain LOCAL_ONE?
```
Agree. We should switch to try-quorum-fallback-to-local for all types of users. But not in this series.

avikivity commented on auth/authenticator.hh:
```diff
@@ -120,7 +128,7 @@ public:
 
-    virtual ::shared_ptr<sasl_challenge> new_sasl_challenge() const = 0;
+    virtual ::shared_ptr<sasl_challenge> new_sasl_challenge(session_dn_func = {}) const = 0;
 };
```
It's wrong to insert a TLS related function into the interface. IIUC SASL isn't tied to TLS in any way. 
As a result password_authenticator receives this session_dn_func which it doesn't know what to do with.

Can't we use it in the concrete class that implements the interface?

elcallio replied:
As stated below, certificates != TLS. The interface could just as well be new_sasl_challenge(std::vector<std::string> dn), but an optional callback saves work when not needed. Point is still just augmenting the sasl call with an option to derive data from certificate info, not making it wire aware. What I could agree with is that we could separate the interfaces and instead of trying to call something called SASL direct from server.cc, have a separately named interface. But it seems a little overzealous also...

avikivity commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,152 @@
+            co_return co_await ca.authenticate(dninfo->subject);
+        }
+    };
+    return ::make_shared<cert_sasl_challenge>(*this, std::move(f));
```
I don't completely understand the details here, but how is this related to sasl challenge/response? We aren't sending a challenge to the client, we're just evaluating already-provided information from TLS.

Why do the dance with pretending to be SASL while in practice we're not?

As a simplification I'd replace create_immediate_sasl_response() with try_authenticate_directly() or something. And transport/server.cc can do the extraction of the TLS information since that's what knows that we're indeed using TLS, while the various authenticators don't know if we're using TLS.

Maybe we don't even need to involve the authenticator. We're bypassing sasl so let's not pretend to be sasl.

elcallio replied:
We need to involve the authenticator to handle resolved subjects (which are just mapping to roles). 
While it is true that cert DN has little to do with SASL, you can argue the session_dn_func argument is optional, and only provided iff sasl request is done via a certificate aware context (not even wire, just something involving certificates - remember that TLS is not same as sockets. And certificates are not same as TLS).

avikivity replied:
But there is clearly a mismatch here between sasl and certificates.

sasl = negotiating proof of authenticity with the client, via a well-defined layered protocol on top of a less-defined rpc protocol (cql here)
certificates = client already proved their identity, we just need to translate to roles

Why even go through the authenticator? Just look up the name via role_manager. Authentication is already done.

elcallio replied:
I guess that could make sense. Instead of `new_sasl_challenge`, simply add an extra call to the auth interface, something akin to the already added `create_immediate_sasl_response`, but call it something like `immediate_dn_auth`. Only downside it having to replicate some logic from `process_auth_response`, but in the end it might look cleaner.

avikivity commented:
Please add a link to dtest. I'm especially interested in negative tests (login should fail).

elcallio commented:
https://github.com/scylladb/scylla-dtest/pull/2939

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4537/

Ten0 commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,152 @@
+        try {
+            std::regex ex(query);
+            if (ex.mark_count() != 1) {
+                throw std::invalid_argument("Role query must have exactly one mark expression");
```
If a user wants to:
- authenticate admins via their emails
- authenticate servers via their CN/SAN

then it looks like what really matters is that there is only one group that matches when resolving the regex (actually even then we could probably take the first one), but there may be several of them defined, e.g. https://regex101.com/r/Slki3N/3

That doesn't seem needed anymore if we make the `auth_certificate_role_query` a list.

elcallio replied:
The check should be `ex.mark_count() == 0`, and we should allow for arbitrary match clauses, picking the first one that fits.

Ten0 reviewed (commented):
Really cool to see this moving forward!
Love that you went with the regex! That's gonna be super simple and flexible! Thanks a lot! ❤️
I noticed a few things while reading, hope that can help!

mykaul commented:
@elcallio - is this going to make it to 5.3 (approx end of April) ?

elcallio commented:
Probably not. I don't think we've even fully decided how this should work. It is a (true) RFC after all.

elcallio commented:
Rebased and changed the cert-based auth to no longer go via sasl-object, but its own early auth. Would like some more discussion/agreements on whether to support SAN stuff before I rework seastar and change the role parsing.

elcallio commented:
Updated again; Now auth-expressions are a list in config, first one to match is used. Requires https://github.com/scylladb/seastar/pull/1629, but can now use alt names as well. (See https://github.com/scylladb/scylla-dtest/pull/2939 tests for some usage). 
Basically:
```
authenticator': 'com.scylladb.auth.CertificateAuthenticator
auth_certificate_role_queries:
   - source: SUBJECT
      query: <regex>
   - source: ALTNAME
      query: <other regex>
```

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1017/

elcallio commented:
ping? opinions welcome.

avikivity commented:
CI is failing. I'll re-review.

avikivity commented on transport/server.cc:
```diff
@@ -897,13 +898,36 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
+                        ss << delim << a;
+                        delim = ",";
+                    }
+                    co_return ss.str();
```
This can be written as: `co_return fmt::format("{}", fmt::join(altnames, ","));`. We're trying to move away from iostreams.

elcallio replied:
Changed.

avikivity commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,181 @@
+#include <regex>
+
+#include "utils/class_registrator.hh"
+#include "replica/database.hh"
```
Probably unnecessary to include this expensive replica-side header.

elcallio replied:
Fixed.

avikivity commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,181 @@
+
+auth::certificate_authenticator::certificate_authenticator(cql3::query_processor& qp, ::service::migration_manager&)
+    : _queries([&] {
+        auto& conf = qp.db().get_config();
```
This db() is data_dictionary, not replica::database. data_dictionary/data_dictionary.hh.

elcallio replied:
See above

avikivity commented:
Looks good. But please add documentation. /cc @annastuchlik .

These days we can run tests involving the client via test.py, but don't know if it's set up for TLS (probably not).

elcallio commented:
Rebased

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1933/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1937/

avikivity commented:
Will re-review.

avikivity commented on transport/server.cc:
```diff
@@ -836,7 +837,16 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
+                throw exceptions::protocol_exception("Immediate (non-password) authentication failed");
+            });
+        } else {
+            res = make_autheticate(stream, a.qualified_java_name(), trace_state);
```
@elcallio ^

avikivity replied:
well it's preexisting, no need to fix it in this series

avikivity commented on auth/certificate_authenticator.cc:
```diff
@@ -0,0 +1,181 @@
+
+                    std::transform(source.begin(), source.end(), source.begin(), ::toupper);
+
+                    std::regex ex(query);
```
We stopped using std::regex due to stalls.

Please switch to boost::regex, and pre-compute the regex so it isn't compiled on every connection.

elcallio replied:
This is the pre-computation. It is literally the loop in constructor transforming config values into runtime matchers. 
Honestly, it seems unlikely also given the kind of regex one expects here, that std:: vs boost:: would yield any measurable difference, but will switch...

avikivity commented:
Looks good, with minor comments.

Missing: test, documentation. Since the documentation is now in-tree operator documentation should come with the patches. @annastuchlik can help.

elcallio commented:
Added tentative docs.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1954/

annastuchlik commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+---------
+
+#. Enable authentication and define authorized roles in the cluster as described in the 
+`Enable Authentication </operating-scylla/security/authentication/>` document. 
```
These changes are required to fix the numbered list and the link (now they don't render properly).

elcallio replied:
fixed

annastuchlik commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+
+#. Enable CQL transport TLS using client certificate verification in each node.
+
+.. code-block:: yaml
+
+   client_encryption_options:
+      enabled: True
+      certificate: <server cert>
+      keyfile: <server key>
+      truststore: <shared trust>
+      require_client_auth: True
+
+Create the client certificates such that each one designates the desired login role for this connection. A typical method 
+would be to assign the subject common name as the role name. All client certificates need to be signed in such a way that 
+either system trust or the defined trust store can verify them.
```
```suggestion
   .. code-block:: yaml

      client_encryption_options:
         enabled: True
         certificate: <server cert>
         keyfile: <server key>
         truststore: <shared trust>
         require_client_auth: True

   Create the client certificates such that each one designates the desired login role for this connection. A typical method 
   would be to assign the subject common name as the role name. All client certificates need to be signed in such a way that 
   either system trust or the defined trust store can verify them.
```

annastuchlik replied:
These changes are required to fix the numbered list.

annastuchlik commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+
+#. Configure the certificate based authenticator in each node.
+
+.. code-block:: yaml
+
+   authenticator: com.scylladb.auth.CertificateAuthenticator
+   auth_certificate_role_queries:
+            - source: SUBJECT
+              query: CN=([^,\s]+)
+
+More than one expression can be used to extract role names. The first one mathing wil be used. The available sources are `SUBJECT` (distinguished name) and
+`ALTNAME` (subject extensions). The query should be a regular expression with a single match clause.
+
+Restart the cluster.
```
```suggestion
   .. code-block:: yaml

      authenticator: com.scylladb.auth.CertificateAuthenticator
      auth_certificate_role_queries:
               - source: SUBJECT
                 query: CN=([^,\s]+)

   More than one expression can be used to extract role names. The first one mathing wil be used. The available sources are `SUBJECT` (distinguished name) and
   `ALTNAME` (subject extensions). The query should be a regular expression with a single match clause.

Restart the cluster.
```

annastuchlik replied:
These changes are required to fix the numbered list

elcallio replied:
Fixed

annastuchlik commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+Certificate Based Authentication
```
```suggestion
Certificate-Based Authentication
```

elcallio replied:
Done

annastuchlik commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+================================
+
+Certificate based authentication is an alternative method of resolving a login to credentials, i.e. a defined ScyllaDB role.
+Instead of defining one or more user/password pairs, each defined role is assigned depending on verified transport 
+client certificate Subject/Alt information.
```
"verified transport 
client certificate Subject/Alt information." - Is that lever of detail required for any reason? Could we just say "verified 
client certificate information"?

elcallio replied:
Perhaps overly specific.

annastuchlik commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+#. Enable authentication and define authorized roles in the cluster as described in the 
+`Enable Authentication </operating-scylla/security/authentication/>` document. 
+
+#. Enable CQL transport TLS using client certificate verification in each node.
```
```suggestion
#. Enable CQL transport TLS using client certificate verification in each node by configuring 
    the `` client_encryption_options`` option in the ``/etc/scylla/scylla.yaml`` file:
```

elcallio replied:
Done

annastuchlik commented on docs/operating-scylla/security/enable-authorization.rst:
```diff
@@ -115,6 +115,23 @@ For example:
 
 .. _roles:
 
+Setting default inital super user and password using config file
```
```suggestion
Setting default initial superuser credentials using the config file
```

elcallio commented on docs/operating-scylla/security/certificate-authentication.rst:
```diff
@@ -0,0 +1,52 @@
+            - source: SUBJECT
+              query: CN=([^,\s]+)
+
+More than one expression can be used to extract role names. The first one mathing wil be used. The available sources are `SUBJECT` (distinguished name) and
```
Borked spelling/sentence. Updated. Should read: "The first matching expression will be used". Fixed.

annastuchlik commented:
@elcallio I've added some suggestions when I was certain how to modify the line and left some questions about the lines I wasn't sure about.

annastuchlik commented on docs/operating-scylla/security/enable-authorization.rst:
```diff
@@ -115,6 +115,23 @@ For example:
+...................................................................
+
+The default super username/password in ScyllaDB is 'cassandra'/'cassandra'. This is very insecure and should be altered using the above CQL procedure. However, to avoid 
+executing for a period before said CQL modifications can be done, the super user name and password can instead be configured in scylla.yaml
```
@elcallio This is the last one :) When it's merged, I'll republish the docs to see if everything renders correctly.

elcallio replied:
Made a slightly different/augmented change. Let me know what you think.

avikivity commented:
What about tests?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2081/

elcallio commented:
> What about tests?

https://github.com/scylladb/scylla-dtest/pull/2939

annastuchlik reviewed (approved):
LGTM
